[TextureMapper] Missing implementation of backingStoreMemoryEstimate

pgorszkowski-igalia opened the pull request:
#### 3ed63da0ea8dcf64cb82bd1a61e79c38a2aeaae0
<pre>
[TextureMapper] Missing implementation of backingStoreMemoryEstimate
<a href="https://bugs.webkit.org/show_bug.cgi?id=254910">https://bugs.webkit.org/show_bug.cgi?id=254910</a>

Reviewed by Miguel Gomez.

The backingStores used by the CoordinatedGraphicsLayers are created during a layerFlush,
and they are freed in the next compositor cycle, after their content is uploaded into
a texture. Due to this, the life span of the Nicosia::Buffers is very short. As the
ResourceUsageThread samples the used memory every 500ms, it would be quite common that
the memory usage reported by the timeline would be zero. In order to avoid this,
the maximum memory usage is stored until the ResourceUsageThread is able to read it.
Thanks to this, the timeline is able to show the peak in the memory usage that reflects
the reality. It&apos;s not perfect though, as the peak will be shown in the timeline some ms
after it really happened, but it&apos;s the best approach we can do.

* Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
(WebCore::ResourceUsageThread::platformSaveStateBeforeStarting):
(WebCore::ResourceUsageThread::platformCollectMemoryData):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::Buffer::resetMemoryUsage):
(Nicosia::Buffer::getMemoryUsage):
(Nicosia::Buffer::Buffer):
(Nicosia::Buffer::~Buffer):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::backingStoreMemoryEstimate const):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:

Canonical link: <a href="https://commits.webkit.org/266424@main">https://commits.webkit.org/266424@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/674c03576275cf3ec33ea23c90091d878e0716f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13829 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14143 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15565 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13138 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13910 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16651 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14225 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15812 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13996 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14616 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11718 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16269 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11903 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12478 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19516 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12979 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12642 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15857 "1 api test failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13172 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11052 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12443 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3358 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16775 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13015 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

magomez commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
```diff
@@ -1524,6 +1526,15 @@ void CoordinatedGraphicsLayer::dumpAdditionalProperties(TextStream& textStream,
+        return 0;
+
+    return *estimatedBackingStoreMemory;
+}
```
You could just return estimatedBackingStoreMemory.value_or(0) to avoid checking whether there value is defined

blino replied:
@pgorszkowski-igalia @magomez : is it normal to return 0 here for `CoordinatedGraphicsLayer::backingStoreMemoryEstimate()`?

It seems that intermediate steps of the PR returned `estimatedBackingStoreMemory`.
cc @hwti

magomez commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
```diff
@@ -253,6 +255,8 @@ class WEBCORE_EXPORT CoordinatedGraphicsLayer : public GraphicsLayer {
     RefPtr<AnimatedBackingStoreHost> m_animatedBackingStoreHost;
     RefPtr<CoordinatedGraphicsLayer> m_backdropLayer;
+
+    std::optional<double> estimatedBackingStoreMemory;
```
I'm not sure why you use an optional here. A double initialized to 0 should be enough, as 0 is what you're returning when the optional is not initialized. I mean, you can use the optional, but it's an overkill in this case.

pgorszkowski-igalia replied:
You are right, I will use just double for that.

magomez commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -965,6 +966,11 @@ bool RenderLayerCompositor::updateCompositingLayers(CompositingUpdateType update
+#if ENABLE(RESOURCE_USAGE)
+        // We save the total memory size of layers to show in the memory timeline of Inspector.
+        WebCore::ResourceUsageThread::setTotalLayerBackingStoreBytes(m_obligatoryBackingStoreBytes + m_secondaryBackingStoreBytes);
+#endif
```
The problem with this is that it's only executed if LOG is enabled, which is odd. So in order to get a valid value we need to enable both LOG and RESOURCE_USAGE. We should be able to get this value without having to enable LOG.

pgorszkowski-igalia replied:
The problem is that m_obligatoryBackingStoreBytes and m_secondaryBackingStoreBytes are set and calculated also only in case when LOG(Compositing) is enabled. 
RESOURCE_USAGE I will remove, it is not needed here.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/MemoryTimelineRecord.js:
```diff
@@ -83,8 +83,10 @@ WI.MemoryTimelineRecord = class MemoryTimelineRecord extends WI.TimelineRecord
-            memoryCategories.push({type: WI.MemoryCategory.Type.Layers, size: layersSize});
+
+        // Memory usage of "Layers" can be "0"
+        memoryCategories.push({type: WI.MemoryCategory.Type.Layers, size: layersSize});
```
what is the motivation/goal of this change?

also, i think this means that _every_ `WI.MemoryTimelineRecord` will now have a `WI.MemoryCategory.Type.Layers` since `layersSize` defaults to `0` at the top, which is probably not what we want.  IMO if a `WI.MemoryTimelineRecord` has a `WI.MemoryCategory.Type.Layers` with a `size` of `0` then we probably dont want/need to show it since there's nothing to show as it is 0 after all (i.e. having an empty graph is arguably worse than not having a graph as it could draw attention where none is deserved)

pgorszkowski-igalia replied:
Actually it is possible that in some cases the memory usage could be 0 but not for all time. And not having graph because memory usage is 0 at the beginning is not ideal solution. Probably there should be something else which will indicate that the graph should not be shown because there will be always 0 value. Maybe some additional field in MemoryCategoryInfo to indicate is it valid or not for specific category? WDYT?
Another solution could be removing categories for which there is no data.

dcrousso replied:
im pretty sure that's how it works today.  it should only show the graph(s) if there's something to show

there's no way for us to know if something in that category will happen in the future.  we certainly could do a better job of differentiating "there was no work in this category" and "the work in this category was not substantial enough to actually be above `0`", but IMO that's something unrelated to this PR so i'd kindly ask that you undo your changes here and open a new bug/PR where we can directly discuss/address that

pgorszkowski-igalia replied:
Sure, I will undo it and create a separate bug report about that. I completely agree with you that this is not related to this PR, but this change will depend on the one about described issue.

pgorszkowski-igalia replied:
https://bugs.webkit.org/show_bug.cgi?id=259553

magomez commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -105,6 +105,9 @@ void ResourceUsageThread::platformSaveStateBeforeStarting()
 #endif
+    // Reset max of reported memory usage of layers.
+    Locker locker { m_layersMemoryUsageLock };
+    m_maxLayersMemoryUsage = 0;
```
this should probably be reset to m_currentLayersMemoryUsage and not 0.

magomez commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -320,6 +323,15 @@ void ResourceUsageThread::platformCollectMemoryData(JSC::VM* vm, ResourceUsageDa
+        Locker locker { m_layersMemoryUsageLock };
+        data.categories[MemoryCategory::Layers].dirtySize = m_maxLayersMemoryUsage;
+        // Reset the max reported of memory usage to gather new max in next 500ms.
+        m_maxLayersMemoryUsage = 0;
```
this should probably be reset to m_currentLayersMemoryUsage and not 0 as well.

magomez reviewed (changes requested):
Please, fix the commit message to explain why this is implemented that way it is, because it's not straightforward to get what's been done here and why.

pgorszkowski-igalia committed:
[TextureMapper] Missing implementation of backingStoreMemoryEstimate
https://bugs.webkit.org/show_bug.cgi?id=254910

Reviewed by Miguel Gomez.

The backingStores used by the CoordinatedGraphicsLayers are created during a layerFlush,
and they are freed in the next compositor cycle, after their content is uploaded into
a texture. Due to this, the life span of the Nicosia::Buffers is very short. As the
ResourceUsageThread samples the used memory every 500ms, it would be quite common that
the memory usage reported by the timeline would be zero. In order to avoid this,
the maximum memory usage is stored until the ResourceUsageThread is able to read it.
Thanks to this, the timeline is able to show the peak in the memory usage that reflects
the reality. It's not perfect though, as the peak will be shown in the timeline some ms
after it really happened, but it's the best approach we can do.

* Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
(WebCore::ResourceUsageThread::platformSaveStateBeforeStarting):
(WebCore::ResourceUsageThread::platformCollectMemoryData):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::Buffer::resetMemoryUsage):
(Nicosia::Buffer::getMemoryUsage):
(Nicosia::Buffer::Buffer):
(Nicosia::Buffer::~Buffer):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::backingStoreMemoryEstimate const):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:

Canonical link: https://commits.webkit.org/266424@main
